Ensure that the response-origin of range requests match the full request (issue 12744)

Snuffleupagus opened the pull request:
The following cases are excluded in the patch:
 - The Firefox PDF Viewer, since it has been fixed on the platform side already; please see https://bugzilla.mozilla.org/show_bug.cgi?id=1683940

 - The `PDFNodeStream`-implementation, used in Node.js environments, since after recent changes that code only supports `file://`-URLs.

Also updates the `read`-methods, on both `PDFNetworkStreamFullRequestReader` and `PDFNetworkStreamRangeRequestReader`, to await the headers before returning any data similarly to the implementation in `src/display/fetch_stream.js`.

*Note:* The relevant unit-tests are updated to await the `headersReady` Promise before dispatching range requests, since that's consistent with the actual usage in the `src/`-folder.

Fixes #12744

Snuffleupagus committed:
Ensure that the response-origin of range requests match the full request (issue 12744)

The following cases are excluded in the patch:
 - The Firefox PDF Viewer, since it has been fixed on the platform side already; please see https://bugzilla.mozilla.org/show_bug.cgi?id=1683940

 - The `PDFNodeStream`-implementation, used in Node.js environments, since after recent changes that code only supports `file://`-URLs.

Also updates the `PDFNetworkStreamFullRequestReader.read`-method to await the headers before returning any data, similar to the implementation in `src/display/fetch_stream.js`.

*Note:* The relevant unit-tests are updated to await the `headersReady` Promise before dispatching range requests, since that's consistent with the actual usage in the `src/`-folder.

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented:
/botio test

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio unittest

Rob--W commented on src/display/fetch_stream.js:
```diff
@@ -121,6 +124,8 @@ class PDFFetchStreamReader {
       createFetchOptions(headers, this._withCredentials, this._abortController)
     )
       .then(response => {
+        stream._responseOrigin.resolve(getResponseOrigin(response.url));
```
This part may not be reached when the network request fails. Is it desirable to indefinitely postpone the resolution of the other responses?

Snuffleupagus replied:
The latest version of the patch updates the unit-tests to agree with the "actual" usage, since that removes the need for the additional Promises in the code.

Rob--W commented on src/display/network_utils.js:
```diff
@@ -36,6 +36,15 @@ function createHeaders(isHttp, httpHeaders) {
+  } catch {
+    // `new URL()` will throw on incorrect data.
+  }
+  return null;
```
```suggestion
  // Notably, null is distinct from "null" string (e.g. from file:-URLs).
  return null;
```

Added extra comment to emphasize the difference. The original issue primarily mattered for http(s):-URLs because these are the only ones supporting range requests, so the exact result for non-http(s) URLs is not super important.

Rob--W commented on src/display/network.js:
```diff
@@ -420,6 +430,14 @@ class PDFNetworkStreamRangeRequestReader {
   }
 
+  _onHeadersReceived() {
+    this.#responseOrigin.resolve(
```
Again this stage may never be reached. Shouldn't we resolve (or reject) earlier to avoid requests getting stuck?

Snuffleupagus replied:
This would be similar to the situation outlined for `src/display/fetch_stream.js` above.

Rob--W reviewed (commented):
The change itself looks good now. Is it possible to add a unit test for this? The test server could blindly redirect to a different origin.

If you don't have two host names for the same server, an easy way to get a cross-origin domain pointing to the same server for free is by appending a dot to the domain, to get a FQDN. These usually resolve to the same server but are considered cross-origin.

E.g. https://example.com/ vs https://example.com./

Rob--W reviewed (changes requested):
> Thanks for the review.
> 
> > Is it possible to add a unit test for this? The test server could blindly redirect to a different origin.
> 
> Unfortunately I don't think I'll have time to work on that any time soon, so hopefully that's not a blocker for landing this.

I created a PR with unit tests at https://github.com/mozilla/pdf.js/pull/19074.

When I open the test locally, http://[::1]:8888/test/unit/unit_test.html?spec=redirects%20block

1 of 3 tests fail: `network > Redirects > redirects blocked if any response is cross-origin, read once`

Your patch currently stores `_storedError`, which is used at the start of `_read()`. But if `_read()` is called immediately, then initially there is no `_storedError` and it will just wait for the response and pass that through. Please fix that.

Snuffleupagus commented:
/botio test

Rob--W commented on src/display/network.js:
```diff
@@ -420,6 +431,22 @@ class PDFNetworkStreamRangeRequestReader {
+    );
+
+    if (responseOrigin !== this._manager._responseOrigin) {
+      this._readCapability.reject(
```
You don't need a separate `_readCapability` promise if you reject all pending `requests` upon setting the error. Here is a patch that you can apply on top of your current patch. It deletes some code, and tweaks `onError` so that it does not overwrite an existing error (because canceling a request is going to generate a new error and we shouldn't replace the error to avoid confusion - perhaps that should be a comment in the code?). It still passes the existing unit tests:

```diff
diff --git a/src/display/network.js b/src/display/network.js
index e4d1264a3..3beb3257b 100644
--- a/src/display/network.js
+++ b/src/display/network.js
@@ -421,7 +421,6 @@ class PDFNetworkStreamRangeRequestReader {
     };
     this._url = manager.url;
     this._requestId = manager.requestRange(begin, end, args);
-    this._readCapability = Promise.withResolvers();
     this._requests = [];
     this._queuedChunk = null;
     this._done = false;
@@ -437,14 +436,11 @@ class PDFNetworkStreamRangeRequestReader {
     );
 
     if (responseOrigin !== this._manager._responseOrigin) {
-      this._readCapability.reject(
-        new Error(
-          `Expected range response-origin "${responseOrigin}" to match "${this._manager._responseOrigin}".`
-        )
+      this._storedError = new Error(
+        `Expected range response-origin "${responseOrigin}" to match "${this._manager._responseOrigin}".`
       );
-      return;
+      this._onError(0);
     }
-    this._readCapability.resolve();
   }
 
   _close() {
@@ -468,8 +464,7 @@ class PDFNetworkStreamRangeRequestReader {
   }
 
   _onError(status) {
-    this._storedError = createResponseStatusError(status, this._url);
-    this._readCapability.reject(this._storedError);
+    this._storedError ??= createResponseStatusError(status, this._url);
     for (const requestCapability of this._requests) {
       requestCapability.reject(this._storedError);
     }
@@ -488,8 +483,6 @@ class PDFNetworkStreamRangeRequestReader {
   }
 
   async read() {
-    await this._readCapability.promise;
-
     if (this._storedError) {
       throw this._storedError;
     }
```

Rob--W reviewed (approved):
The functional behavior of the patch looks good to me. I added a comment below with a suggestion to simplify, basically getting rid of the new `_readCapability` promise in favor of rejecting existing request promises. Consider this patch approved and ready to merge with that suggestion applied.

A note to a comment in the PR text / commit message:

> Note: The relevant unit-tests are updated to await the headersReady Promise before dispatching range requests, since that's consistent with the actual usage in the src/-folder.

Note: this test fixup was needed because this patch introduces a new requirement: `getRangeReader` can only be invoked on streams that were preceded by a full request, at least until the point that the headers were received. Independently of whether the current code satisfies that requirement, we should document this expectation.

I confirmed your claim that the current code in `src/` awaits `headersReady` before dispatching range requests, with the entry point being at [`getPdfManager` in worker.js](https://github.com/mozilla/pdf.js/blob/1f6cc8513456d7a4fe61a48c6d0a11e7c7ee49fb/src/core/worker.js#L239-L250) - our implementation does this because in order to know whether we need to switch to range requests, we first need to know whether the server supports range requests. Currently the only supported way to do so is to send another request.

As an example of a differing implementation: the Chrome extension has already received the headers at the time that the viewer is opened, so in theory it could forward the headers and response origin to the viewer so that it can immediately try switch to range requests faster (without doing the initial full request). I currently have no plans to implement that in the Chrome extension due to added complexity though.

Snuffleupagus commented:
> Consider this patch approved and ready to merge with that suggestion applied.

Fixed, thank you.

> `getRangeReader` can only be invoked on streams that were preceded by a full request, at least until the point that the headers were received. Independently of whether the current code satisfies that requirement, we should document this expectation.

I've added a comment in `src/interfaces.js`, since that's what all stream implementations are based upon.

> As an example of a differing implementation: the Chrome extension has already received the headers at the time that the viewer is opened, so in theory it could forward the headers and response origin to the viewer so that it can immediately try switch to range requests faster (without doing the initial full request). I currently have no plans to implement that in the Chrome extension due to added complexity though.

So basically, the Chrome extension could be changed to instead use `PDFDataRangeTransport` (similar to the Firefox PDF Viewer).
Although, the part about avoiding additional complexity makes sense.

Snuffleupagus commented:
/botio unittest
